.github/workflows/ci.yml

```diff
@@ -872,12 +872,6 @@ jobs:
     strategy:
       matrix:
         target:
-          - platform: android
-            arch: armv8.2-a
-            abi: arm64-v8a
-            docker_image: "gcr.io/iree-oss/android@sha256:3f641d25786b1e5e430ee4cacb8bfe57540fda5ecaa7ca2802c179c26e77ce09"
-            build_script: "./build_tools/cmake/build_android.sh"
-            # No test_script
```

Comment:
> It breaks out the Android target from the matrix cross_compile_and_test. I think it makes sense to do so as Android has its own matrix to run tests on multiple devices.

I'd prefer to add to the existing matrix (android_pixel_4, android_pixel_6, etc.). We can add another setup script or more env vars as needed for that. This PR is adding a lot of boilerplate code right now.

Reply:
AFAIK we are not able to do that because android test workflow needs two different machines: 1. build machine to do cross-compile and 2. mobile phones to run tests. So they must be run in different jobs, but the matrix only supports a single job.

Ideally in the test step we can call a reusable workflow to run tests on different devices, but I don't think it's possible right now (reusable workflow must be called on the job level: https://docs.github.com/en/actions/using-workflows/reusing-workflows#calling-a-reusable-workflow)

Reply:
Another way is having a matrix job to cross-compile and upload test artifacts, and having another matrix job to download and run tests on different machines. But that introduces unnecessary deps (android tests need to wait for all cross-compilation finished because it can't just wait on the android cross-compile in the matrix)

Reply:
> because android test workflow needs two different machines: 1. build machine to do cross-compile and 2. mobile phones to run tests

Isn't that the point of cross compiling in general though? Every job in the matrix will want to
1. run the already built compiler the generate program artifacts
2. build the runtime (fast, so doesn't need a large build machine)
3. start an emulator, connect to a device, etc.
4. run tests on the emulated platform / connected device

Reply:
Correct, but previously we can run build and test in a single job because we only use emulators to run tests on the same build machine. However it's not the case now because android tests are running on mobile phones and those devices don't connect to the build machine.

So we want something like
```yml
jobs:
  cross_compile_and_test:
    matrix:
      - arch: a
        test_device: pixel-4
      - arch: b
        test_device: riscv_emulator
   steps:
     cross_compile:
       run-on: x86_64 build_machine
       runs: ./cross_compile.sh ${matrix.arch}
    test:
      run-on: ${matrix.test_device}
      runs: ./run_test.sh
```

But the problem is we can't run step on different devices in a single job and the matrix is limited to a single job.

Reply:
Oh, are the Android phones themselves registered as runners, not Linux/Raspberry Pi devices that talk to the phones over ADB?

Reply:
This is getting into limitations in what GitHub gives us and raising the question for me again about whether we want to generate our workflows...

Another Python framework for generating JSON!

Reply:
> 2. build the runtime (fast, so doesn't need a large build machine)

Since mobile phones connect to RPi, I don't think it's really feasible to build runtime on RPi, even though it's a lightweight job for x86 machine (also the pre-built host tools don't run on arm_64)

Reply:
> Oh, are the Android phones themselves registered as runners, not Linux/Raspberry Pi devices that talk to the phones over ADB?

The Android phones connect to RPi, and RPi are registered as runners. But still I feel it's not easy/slow to cross-compile runtime for Android on the RPi (you need Android NDK and the pre-built iree-compile don't run on ARM64 RPi)

Reply:
Brainstorming:

* Could we run tests on Android emulators instead of physical devices?
* Could we have one job that builds for each cross compile target (or multiple: `crosscompile_host_linux`, `crosscompile_host_macos`), then a matrix that is just control runners (e.g. RPis, machines running emulators) that use those already built artifacts?
* Could we use Linux machines instead of RPis to connect to the physical Android phones?

I'm worried about CI overheads (compress -> upload -> wait for runner -> download -> decompress) and workflow complexity. If we add iOS devices, embedded systems (e.g. Arduino), or more Android phones, I'd like for them to show up in the same part of the actions graph and be easy to edit in the workflow file: 
![image](https://user-images.githubusercontent.com/4010439/231823990-6bb9181d-c02a-449c-ab22-6ee5a1cd9692.png)

Reply:
> Could we run tests on Android emulators instead of physical devices?

This might be possible for some CPU tests, but we also run GPU tests that need physical devices, AFAIK

> Could we have one job that builds for each cross compile target (or multiple: crosscompile_host_linux, crosscompile_host_macos), then a matrix that is just control runners (e.g. RPis, machines running emulators) that use those already built artifacts?

This is a possible solution. But as I mentioned above, we will force all jobs in the "test matrix" to wait on all jobs in the "cross-compile matrix", because GH doesn't provide fine-grained dependency control between matrix (e.g. Your job can't only wait on a single job in the matrix). I'm fine with this route if we are okay with unnecessary dependencies between jobs.

We also need  some tricks to pass output between matrix jobs (e.g. GCS url of the cross-compiled artifacts). See
https://github.com/openxla/iree/blob/586f19e0f6c9d9e76200498af855905820cbd390/.github/workflows/build_benchmark_tools.yml#L53-L60
(and although Github is working on a feature to store all outputs from every job in a matrix, the feature seems to be delayed https://github.com/actions/runner/pull/2477#issuecomment-1501003600)

> Could we use Linux machines instead of RPis to connect to the physical Android phones?

Might be. But I don't feel our lab has enough resources right now (given the previous experience to set up other devices). 

In terms of the overhead to upload and download. The archive + upload is quite fast (14s) but download + unarchive is slow (7m). I think it's due to the slow SD card speed and CPU performance on RPi (but unless we stop using RPi, I don't think cross-compile on RPi will be faster). The total size of archive is 1.6GB but we can remove `*.o` and `*.a` to cut it down to 1.3GB. The test binary is actually surprisingly large to me. Like each `runtime/src/iree/hal/drivers/local_task/cts/local-task_vmvx_*` test binary is 12M.

Reply:
> In terms of the overhead to upload and download, the archive + upload is quite fast (14s) but download + unarchive is slow (7m).

That's pretty slow for pure overhead of something we want to run on presubmit. Will that be a bottleneck if we have multiple PRs and merged commits all queuing to use a single RPi? We might be able to proceed anyways, but the time burnt on overhead and the extra workflow complexity concerns me.

> > Could we run tests on Android emulators instead of physical devices?
> 
> This might be possible for some CPU tests, but we also run GPU tests that need physical devices, AFAIK

Desktop GPU machines should be plenty powerful enough to build the runtime and compile any artifacts they need though. Mobile GPU tests on the other hand will want physical devices, yeah (probably not much use testing the Android Vulkan loader + SwiftShader when we already test SwiftShader on desktop).

Reply:
> Desktop GPU machines should be plenty powerful enough to build the runtime and compile any artifacts they need though

? We're only talking about Android here. We definitely would like to run Android emulator tests, but we're still going to need physical devices for real GPU tests and benchmarking.

It's annoying that GitHub Actions doesn't give us the right tools here. Looking at the graph, it seems like making the Android tests depend on the full cross-compilation wouldn't be too awful:

![image](https://user-images.githubusercontent.com/5732088/232574898-0b9aa358-ce2f-4d0e-a2f6-a1335fb01f98.png)

If we do, we should definitely leave a note about what's going on though.

An alternative would be to factor cross-compile into a reusable workflow, so that there's less duplication. I'm not sure whether passing the all the arguments ends up being as complicated as just copy-pasting the whole thing though.

Reply:
If we put it in a reusable workflow, can we get a matrix to run multiple jobs? https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-a-matrix-strategy-with-a-reusable-workflow

The reusable workflow would do both compilation and testing and that would be abstracted away from the caller

Reply:
After trying more with reusable workflows and matrix jobs, I decided to stick with the original separate android workflow. Added the comment to explain why: https://github.com/openxla/iree/pull/13024/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR935-R941

Basically the requirements of running on non-scalable physical devices and limited capabilities of matrix jobs make the Android test special. Eventually other platforms might also have physical devices to run, at that point we can generalize the current Android workflow to reuse it. But that will require more customisable/dynamic logic in the workflow and I don't think it will happen in the near future, so maybe we shouldn't over-optimized the solution for now

Reply:
Yeah, while I'm not thrilled that we have to do it this way, the other near-term options are not awesome either. I think we should merge this as-is